Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,308 @@
+            var parameters = new BindingParameterCollection();
+            parameters.Add(optnVal);
+            parameters.Add(_services);
+            if (_customBinding.CanBuildServiceDispatcher<IInputChannel>(parameters))
+            {
+                _serviceDispatcher = _customBinding.BuildServiceDispatcher<IInputChannel>(parameters, dispatcher);
+            }
+            parameters.Add(_serviceDispatcher);
+            BindingContext bindingContext = new BindingContext(_customBinding, parameters);
+            QueueTransportPump queuePump = queueTransportBinding.BuildQueueTransportPump(bindingContext);
+
+            _queueTransportContexts.Add(new QueueTransportContext
+            {
+                QueuePump = queuePump,
+                 ServiceDispatcher = _serviceDispatcher,
+                 QueueBindingElement = queueTransportBinding,
+                 MessageEncoderFactory = msgEncBindingelement.CreateMessageEncoderFactory(),
+                 QueueHandShakeDelegate = _queueHandShakeMiddleWare.Build()
+            });
+
+        }
+    }
+
+}
+```
+- for each service type (which is fetched from DI), service dispatcher is created.
+- for each service dispatcher, validation is made to ensure the service dispatcher binding element collection property has a transport binding element of type `QueueBaseTransportBindingElement`.
+- using the `QueueBaseTransportBindingElement` which is fetched from above step, `BuildQueueTransportPump` method is called to get the `QueueTransportPump`.
+ - In the end, for every service dispatcher, a `QueueTransportContext` object is created which stores objects of type `QueueTransportPump`, `ServiceDispatcher`, `QueueBaseTransportBindingElement`,`MessageEncoderFactory` and `QueueHandShakeDelegate`.
+- Since `QueuePollingService` implements `IHostingService`, it implements the `StartAsync` and `StopAsync` methods.
+- `StartAsync` is called by WebHost and it performs the below task.
+```csharp
+public async Task StartAsync(CancellationToken cancellationToken)
+{
+    var tasks = _queueTransportContexts.Select(_queueTransport => StartFetchingMessage(_queueTransport));
+    await Task.WhenAll(tasks);
+}
+
+private async Task StartFetchingMessage(QueueTransportContext queueTransport)
+{
+    await queueTransport.QueuePump.StartPumpAsync(queueTransport, CancellationToken.None);
+}
+
+```
+- For StartAsync
+ - First we create a list of tasks`StartFetchingMessage` for each `QueueTransportContext` (populated during Init).
+ - `QueueTransportPump` from `QueueTransportContext` is used by each `StartFetchingMessage` to call the `StartPumpAsync` method.
+ - If the `QueueTransportPump` is based on a push message mechanism, it awaits until a message is received and invokes service as soon as the message is received.
+ - If the `QueueTransportPump` is using `DefaultQueueTransportPump`, it creates multiple concurrent tasks to invoke `FetchAndProcessAsync` method based on concurrency level.
```

Comment:
If the `QueueTransportPump` is based on a push message mechanism, the provided implementation awaits...  

For push, the implementer needs to provide a concrete implementation instead of proving an in-box implementation. Without this small change in wording, I think it might come across as there's one in box which does this already.

Reply:
I also think saying "awaits" implies an implementation which might not be the case. With the EventHandler mechanism some libraries use you aren't awaiting anything. Maybe replace awaits with waits as that's more generic, and awaiting is a type of waiting, as is an EventHandler being executed. I know that's a lot of words to say awaits -> waits, but wanted to give my rationale.